File: js/storage/activities.js
/**
 * Activity storage — IndexedDB backed.
 * Stores completed ride data.
 */

const DB_NAME = 'fz_activities';
const DB_VERSION = 1;
const STORE_NAME = 'activities';

function openDB() {
    return new Promise((resolve, reject) => {
        const req = indexedDB.open(DB_NAME, DB_VERSION);
        req.onupgradeneeded = (e) => {
            const db = e.target.result;
            if (!db.objectStoreNames.contains(STORE_NAME)) {
                const store = db.createObjectStore(STORE_NAME, { keyPath: 'id' });
                store.createIndex('date', 'date', { unique: false });
            }
        };
        req.onsuccess = () => resolve(req.result);
        req.onerror = () => reject(req.error);
    });
}

/**
 * Save a completed ride activity.
 * @param {Object} data - Ride data payload
 * @returns {string} Activity ID
 */
export async function saveActivity(data) {
    const now = new Date();
    const id = now.toISOString().replace(/[-:T.]/g, '').slice(0, 15);

    const activity = {
        id,
        date: now.toISOString(),
        route_name: data.route_name || 'Unknown',
        duration_s: Math.round((data.duration_s || 0) * 10) / 10,
        distance_m: Math.round((data.distance_m || 0) * 10) / 10,
        avg_power_w: Math.round((data.avg_power_w || 0) * 10) / 10,
        max_power_w: Math.round(data.max_power_w || 0),
        avg_cadence: Math.round((data.avg_cadence || 0) * 10) / 10,
        avg_speed_kmh: Math.round((data.avg_speed_kmh || 0) * 10) / 10,
        elevation_gain_m: Math.round((data.elevation_gain_m || 0) * 10) / 10,
        power_samples: data.power_samples || [],
        track_samples: data.track_samples || [],
    };

    const db = await openDB();
    const tx = db.transaction(STORE_NAME, 'readwrite');
    tx.objectStore(STORE_NAME).put(activity);

    await new Promise((resolve, reject) => {
        tx.oncomplete = resolve;
        tx.onerror = () => reject(tx.error);
    });

    console.log(`[ACTIVITY] Saved: ${id}`);
    return id;
}

/**
 * List all activities, newest first.
 */
export async function listActivities() {
    const db = await openDB();
    const tx = db.transaction(STORE_NAME, 'readonly');

    return new Promise((resolve, reject) => {
        const req = tx.objectStore(STORE_NAME).getAll();
        req.onsuccess = () => {
            const activities = req.result
                .map(a => {
                    // Strip large arrays for list view
                    const { power_samples, track_samples, ...rest } = a;
                    return rest;
                })
                .sort((a, b) => b.date.localeCompare(a.date));
            resolve(activities);
        };
        req.onerror = () => reject(req.error);
    });
}

/**
 * Get a single activity by ID (with full data).
 */
export async function getActivity(id) {
    const db = await openDB();
    const tx = db.transaction(STORE_NAME, 'readonly');
    return new Promise((resolve, reject) => {
        const req = tx.objectStore(STORE_NAME).get(id);
        req.onsuccess = () => resolve(req.result || null);
        req.onerror = () => reject(req.error);
    });
}

/**
 * Delete an activity by ID.
 */
export async function deleteActivity(id) {
    const db = await openDB();
    const tx = db.transaction(STORE_NAME, 'readwrite');
    tx.objectStore(STORE_NAME).delete(id);
    await new Promise((resolve, reject) => {
        tx.oncomplete = resolve;
        tx.onerror = () => reject(tx.error);
    });
}

File: UNCHAINED_PROJECT/unchained_project/storage/activities.py
"""Activity persistence — save and list completed rides as JSON files."""

from __future__ import annotations

import json
from datetime import datetime
from pathlib import Path


def save_activity(data: dict, activities_dir: Path) -> str:
    """Save a ride summary to activities/<timestamp>.json.

    Returns:
        Filename of the saved activity.
    """
    activities_dir.mkdir(parents=True, exist_ok=True)

    ts = datetime.now()
    filename = ts.strftime("%Y%m%d_%H%M%S") + ".json"

    activity = {
        "date": ts.isoformat(),
        "route_name": data.get("route_name", "Unknown"),
        "duration_s": round(data.get("duration_s", 0), 1),
        "distance_m": round(data.get("distance_m", 0), 1),
        "avg_power_w": round(data.get("avg_power_w", 0), 1),
        "max_power_w": round(data.get("max_power_w", 0), 1),
        "avg_cadence": round(data.get("avg_cadence", 0), 1),
        "avg_speed_kmh": round(data.get("avg_speed_kmh", 0), 1),
        "elevation_gain_m": round(data.get("elevation_gain_m", 0), 1),
        "power_samples": data.get("power_samples", []),
        "gpx_file": data.get("gpx_file"),
        "strava": data.get("strava"),
    }

    filepath = activities_dir / filename
    with open(filepath, "w") as f:
        json.dump(activity, f, indent=2)

    print(f"  [ACTIVITY] Saved: {filename}")
    return filename


def list_activities(activities_dir: Path) -> list[dict]:
    """Return all saved activities, newest first."""
    if not activities_dir.exists():
        return []

    activities = []
    for fp in sorted(activities_dir.glob("*.json"), reverse=True):
        try:
            with open(fp) as f:
                act = json.load(f)
            act.pop("power_samples", None)
            act["filename"] = fp.name
            activities.append(act)
        except (json.JSONDecodeError, OSError):
            continue

    return activities
